### PR TITLE
Adding background and interface design considerations as agreed upon by Ansible Roles metateam

### PR DESCRIPTION
**Pull request by pcahyna**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,6 +1,20 @@
 This document is intended to be a standards document for the Ansible roles contained in the
 Oasis Roles repositories.
 
+Background
+==========
+
+The goal of the Ansible Metateam project (specifically, the [Linux System Roles project](https://github.com/linux-system-roles)) is to provide a stable and consistent user interface to multiple operating systems (multiple versions of RHEL in the downstream RHEL System Roles package, additionally CentOS, Fedora at least). Stable and consistent means that the same Ansible playbook will be usable to manage the equivalent functionality in the supported versions without the administrator (tghe user of the role) being forced to change anything in the playbook (the roles should serve as abstractions to shield the administrator from differences). Of course, this means that the interface of the roles should be itself stable (i.e. changing only in a backward compatible way). This implies a great  responsibility in the design of the interface, because the interface, unlike the underlying implementation, can not be easily changed.
```
Comment (greg-hellings) [inline README.md]: Typo `tghe`

File: README.md [hunk 2]
```diff
@@ -9,15 +23,67 @@ Basics
...
+
+When there are multiple implementations of the same functionality, we call them “providers”. A role
+supporting multiple providers should have an input variable called `$ROLENAME_provider`. If this
+variable is not defined, the role should detect the currently running provider on the system, and
+respect it. (Rationale: users can be surprised if the role changes the provider if they are running
+one already.) If there is no provider currently running, the role should select one according to the
+OS version. (E.g. on RHEL 7, chrony should be selected as the provider of time synchronization,
+unless there is ntpd already running on the system, or user requests it specifically. Chrony should
+be chosen on RHEL 8 as well, because it is the only provider available.) The role should set a
+variable or custom fact called `$ROLENAME_provider_os_default` to the appropriate default value for
```
Comment (greg-hellings) [inline README.md:105]: What if a role has multiple providers (e.g. one to determine the webserver daemon, and another to determine the JVM)? What's the naming scheme, then?

File: README.md [hunk 3]
```diff
@@ -1,6 +1,20 @@
...
+  * Users that want to manage multiple major releases of the system simultaneously with a single playbook.
+  * Users that want to migrate to a new version of the system without changing their automation (playbook).
+  * Users who want to switch to a different provider in the same version of the OS (like switching from ntpd to chrony to RHEL 7) and keep the same playbook.
+
+Designing the interface in the latter case is difficult because it has to be sufficiently abstract to cover different providers. We, for example, do not provide an email role in the Linux System Roles project, only a postfix role, because the underlying implementations (sendmail, postfix) were deemed to be too divergent. Generally, an abstract interface should be something that should be always aimed for though, especially if there are multiple providers in use already, and in particular when the default provider is changing or is known to be likely to change in the next major releases.
```
Comment (greg-hellings) [inline README.md]: This whole section needs more line breaks. This is going to be difficult to edit in most editors.

**Review by greg-hellings — CHANGES_REQUESTED** [review]

File: README.md (hunk 1 above)
Comment (pcahyna) [reply README.md]: fixed, together with some other typos

File: README.md (hunk 3 above)
Comment (pcahyna) [reply README.md]: fixed

File: README.md (hunk 2 above)
Comment (pcahyna) [reply README.md:105]: We have not discussed this, but the timesync role provides an example: the provider variable for the ntp functionality is called `timesync_ntp_provider`, leaving open the possibility of adding `timesync_somethingelse_provider`: https://github.com/linux-system-roles/timesync/blob/f1a23686dc1107520a18b41c5d1fc52ee60d5fc5/defaults/main.yml#L7

**Review by greg-hellings — APPROVED** [review]

**Review by robled — APPROVED** [review]

**Pull request merged**